httpload/cmd/httpload/main.go

```diff
@@ -0,0 +1,440 @@
+	urlc, errc, err := config.genRequests(ctx)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	resultc, errc, err := lt.genLoadRequests(ctx, config, urlc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	errc, err = lt.genResults(ctx, resultc)
+	if err != nil {
+		return err
+	}
+	errcList = append(errcList, errc)
+
+	err = lt.waitForPipeline(errcList...)
+
+	done <- true
+
+	return err
+}
+
+func (lt *LoadTester) dumpStatuses(w io.Writer) {
+	statuses := make(map[int]int)
+
+	for _, result := range lt.results {
+		statuses[result.status]++
+	}
+
+	keys := make([]int, 0, len(statuses))
+
+	for key := range statuses {
+		keys = append(keys, key)
+	}
+
+	sort.Ints(keys)
+
+	for _, key := range keys {
+		code := float64(key)
+		count := float64(statuses[key])
+		fmt.Fprintf(w, "%s\n", fmt.Sprintf("[%3.f] %12.f response(s)", code, count))
+	}
+}
+
+func (lt *LoadTester) dumpTimings(w io.Writer) {
+	totalTiming := float64(0)
+	maxTiming := float64(0)
+	minTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(lt.results)
+	noResults := (n == 0)
+
+	if noResults {
+		return
+	}
+
+	timing0 := lt.results[0].timing.Seconds()
+	totalTiming = timing0
+	maxTiming = timing0
+	minTiming = timing0
+
+	for i := 1; i < n; i++ {
+		timing := lt.results[i].timing.Seconds()
+
+		totalTiming += timing
+
+		if timing > maxTiming {
+			maxTiming = timing
+		}
+
+		if timing < minTiming {
+			minTiming = timing
+		}
+	}
+
+	averageTiming = totalTiming / float64(n)
+	requestsPerSecond = float64(1) / averageTiming
+
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Total:        %12.4f secs", totalTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", maxTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", minTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(w, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+}
+
+func setupGracefulShutdown(done chan bool) {
+	signalChannel := make(chan os.Signal, 1)
+
+	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGKILL)
+
+	<-signalChannel
+
+	done <- true
+}
+
+func trueMain(flags *flag.FlagSet, args []string) int {
+	done := make(chan bool)
+	defer close(done)
+	go setupGracefulShutdown(done)
+
+	config := newConfigFromFlags(flags)
+
+	if err := config.parse(flags, args); err != nil {
+		fmt.Println("main: error during arguments parsing. Error:", err)
+		flags.Usage()
+		return exitCodeError
+	}
+
+	if err := config.checkFlags(); err != nil {
+		log.Println("main: error checking flags. Error:", err)
+		return exitCodeError
+	}
+
+	loadTester := newLoadTesterFromConfig(config)
+
+	go func() {
+		err := loadTester.run(config, done)
+
+		if errors.Is(err, context.DeadlineExceeded) {
+			return
+		}
+
+		if err != nil {
+			log.Println("main: error during load test. Error:", err)
+			os.Exit(exitCodeError)
+		}
+	}()
+
+	<-done
```

Comment:
We're starting a goroutine and immediately after, we wait for it to be terminated before going on with the program, which means this code could be simpler if `run` was a blocking call.
Generally, it's better to write, reason about and maintain if functions/methods are blocking. This has also the advantage that concurrency constructs such as channels do not leak a function/method scope.

For example: 

```go
...
loadTester := newLoadTesterFromConfig(config)
err := loadTester.run()
if err != nil {
    // handle actual error
    ...   
}
...
```

Reply:
In reality, I can come out with a message on `done` channel in case of shutdown, either. If the call is blocking, I cannot shutdown properly (unless Go magically handles it in a way I do not know). Or do you mean something else?

Reply:
Maybe, I can call the channel `shutdown` or something similar to clarify its function. What do you think?

Reply:
The graceful shutdown was not a requirement for this application, I think it's safe to assume that if you CTRL-C while the loadtester is running you're not interested by the probably incomplete results you'll get. I mean it can  be an improvement sure, but let's add it further.

If we want to keep it, we could remove the done channel and use a context that manages the whole application lifetime.
It gets cancelled on CTRL-C. Also if the config specified a timeout, you create a child context out of it (using `context.WithTimeout(parent, timeout)`.

Anyway, about making the RunLoaderPipeline method blocking, I don't see what you mean, the code is already blocking since the execution stops just after calling RunLoaderPipeline and waits for the done channel.